R/propensity-weighting.R

```diff
@@ -0,0 +1,95 @@
+#' Calculates inverse probability of treatment weights using a random forest
+#'
+#' @param df <`tbl`> Data frame on which to perform weighting
+#' @param treatment <`character(1)`> String with name of treatment column
+#' @param ivs <`character`> Character vector with names of independent variabes to use in propensity score model
+#' @param impute_missing <`lgl`> If `TRUE` will internally impute missing values in IVs for propensity score fit
+#'
+#' @return <`tbl`> Data frame with new column for propensity weights
+#'
+#' @import purrr
+#' @importFrom rlang := !! sym
+#'
+#' @export
+add_propensity_weights <- function(df, treatment, ivs, impute_missing = FALSE) {
+
+  assert_that(is.data.frame(df))
+
+  assert_that(is.string(treatment))
+  assert_that(is_categorical_variable(df, treatment))
+  assert_that(noNA(df[[treatment]]))
+
+  assert_that(is.character(ivs))
+  assert_that(all(ivs %in% colnames(df)))
+
+  fit_df <- df
+
+  missing_ivs <- fit_df %>%
+    select(ivs) %>%
+    is.na() %>%
+    sum()
+
+  if (missing_ivs > 0 & !impute_missing) {
+
+    stop(
+      "\nOne or more IVs have missing data, you should impute or drop missing cases prior to using this function.\n",
+      "If you want to leave the missing data but still calculate propensity scores for every observation, then set `impute_missing = TRUE`.",
+      call. = FALSE
+    )
+
+  }
+
+  if (missing_ivs > 0 & length(ivs) == 1) {
+
+    if (is_categorical_variable(fit_df, ivs)) {
+
+      iv_quo <- sym(ivs)
+
+      most_common_category <- fit_df %>%
+        select(!!iv_quo) %>%
+        drop_na() %>%
+        count(!!iv_quo) %>%
+        arrange(desc(n)) %>%
+        slice(1) %>%
+        pull(!!iv_quo)
+
+      fit_df <- fit_df %>%
+        mutate(!!ivs := replace_na(!!ivs, most_common_category))
+
+    } else {
+
+      median_value <- median(fit_df[[ivs]], na.rm = TRUE)
+
+      fit_df <- fit_df %>%
+        mutate(!!ivs := replace_na(!!ivs, median_value))
+
+    }
+  }
+
+  if (missing_ivs > 0 & length(ivs) > 1) {
+
+      fit_df <- fit_df %>%
+        select(ivs) %>%
+        mutate_if(is.character, as.factor) %>%
+        mice::mice(m = 1, print = FALSE) %>%
```

Comment:
Is there a reason you changed this from the default `m = 5`?  I don't know if it makes much of a difference, but just wondering

Reply:
`m=5` would do multiple imputation with 5 output dataframes. For our situation, single imputation seems good enough. The other option is to do multiple imputation, calculate propensity scores for each imputed set, and average the results.

Reply:
If you run `mice::complete()` after multiple imputation, you still get back one dataframe, which the propensity scores could be calculated from.

Reply:
`mice::complete()` defaults to the first imputed set `mice::complete(action = 1)`. The behavior would be the same, but it would have to impute 4 dataframes that aren't used.

Reply:
I see.  It wasn't clear to me what the default action was there.  Makes sense to just use `m = 1` then.